internal/tasks/account_deletion.go

```diff
@@ -0,0 +1,213 @@
+				Name: "keppel_account_deletion",
+				Help: "Counter for deleted accounts.",
+			},
+		},
+		DiscoverTask: j.discoverAccountForDeletion,
+		ProcessTask:  j.deleteMarkedAccount,
+	}).Setup(registerer)
+}
+
+var (
+	accountDeletionSelectQuery = sqlext.SimplifyWhitespace(`
+		SELECT name FROM accounts
+		WHERE is_deleting AND next_deletion_attempt_at < $1
+		ORDER BY next_deletion_attempt_at ASC, name ASC
+	`)
+)
+
+func (j *Janitor) discoverAccountForDeletion(_ context.Context, _ prometheus.Labels) (accountName models.AccountName, err error) {
+	err = j.db.SelectOne(&accountName, accountDeletionSelectQuery, j.timeNow())
+	return accountName, err
+}
+
+func (j *Janitor) deleteMarkedAccount(ctx context.Context, accountName models.AccountName, labels prometheus.Labels) error {
+	accountModel, err := keppel.FindAccount(j.db, accountName)
+	if err != nil {
+		return err
+	}
+	if errors.Is(err, sql.ErrNoRows) {
+		// assume the account got already deleted
+		return nil
+	}
+
+	actx := keppel.AuditContext{
+		UserIdentity: janitorUserIdentity{TaskName: "account-deletion"},
+		Request:      janitorDummyRequest,
+	}
+	err = j.DeleteAccount(ctx, *accountModel, actx)
+	if err != nil {
+		return fmt.Errorf("could not delete account %q: %w", accountName, err)
+	}
+	return nil
+}
+
+var (
+	deleteAccountFindManifestsQuery = sqlext.SimplifyWhitespace(`
+		SELECT r.name, m.digest
+			FROM manifests m
+			JOIN repos r ON m.repo_id = r.id
+			JOIN accounts a ON a.name = r.account_name
+			LEFT OUTER JOIN manifest_manifest_refs mmr ON mmr.repo_id = r.id AND m.digest = mmr.child_digest
+    WHERE a.name = $1 AND parent_digest IS NULL
+	`)
+	deleteAccountReposQuery                   = `DELETE FROM repos WHERE account_name = $1`
+	deleteAccountCountBlobsQuery              = `SELECT COUNT(id) FROM blobs WHERE account_name = $1`
+	deleteAccountScheduleBlobSweepQuery       = `UPDATE accounts SET next_blob_sweep_at = $2 WHERE name = $1`
+	deleteAccountMarkAllBlobsForDeletionQuery = `UPDATE blobs SET can_be_deleted_at = $2 WHERE account_name = $1`
+)
+
+type deleteAccountRemainingManifest struct {
+	RepositoryName string
+	Digest         string
+}
+
+func (j *Janitor) DeleteAccount(ctx context.Context, account models.Account, actx keppel.AuditContext) error {
+	// can only delete account when user has deleted all manifests from it
+	var nextManifests []deleteAccountRemainingManifest
+	err := sqlext.ForeachRow(j.db, deleteAccountFindManifestsQuery, []any{account.Name},
+		func(rows *sql.Rows) error {
+			var m deleteAccountRemainingManifest
+			err := rows.Scan(&m.RepositoryName, &m.Digest)
+			nextManifests = append(nextManifests, m)
+			return err
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	if len(nextManifests) > 0 {
+		// deletion was not completed yet -> check if we need to do something on our side
+		logg.Info("cleaning up account %q: need to delete %d manifests in this cycle", account.Name, len(nextManifests))
+		for _, rm := range nextManifests {
```

Comment:
This loop can be inlined with the previous one. Instead of collecting into `nextManifests`, each manifest can be deleted immediately.